### ISSUE TITLE
fastboot is unable to handle file greater than 100M?

### ISSUE THREAD

**Issue by MengJueM** [google/python-adb#23]
Here is the error code:

```
fastboot.FastbootRemoteFailure: ('FAIL: data too large', 'data too large')
```

**Comment by fahhem** [comment]
Yes, that feature started coming out when I wrote this. Unfortunately, as far as I can tell, there's a lot of code in C++ to generate the files. If you can figure out how to get the data in 'sparse' format from the image in Python, then I can work on adding this to fastboot. Sorry about that!

**Comment by MengJueM** [comment]
@fahhem Thanks for your comment.

I don't know fastboot protocol very well. Can the big file be downloaded by multiple chunks and then flash once? If yes, the solution will be simple.

**Comment by MengJueM** [comment]
```
    def download(self, source_file):
        file_size = os.stat(source_file).st_size
        if file_size > MAX_ACCEPT_SIZE:
            chunks = [MAX_ACCEPT_SIZE] * (file_size / MAX_ACCEPT_SIZE) + \
                     [file_size % MAX_ACCEPT_SIZE]
            offset = 0
            for i in chunks:
                f = open(source_file)
                f.seek(offset)
                temp = StringIO(f.read(i))
                self.fb_inst.Download(temp, info_cb=self.info_cb)
                offset += i
```

I tried this codes and the big files can be downloaded with returning "OK". I am not sure whether I can flash the files. Yet no idea how to combine the small size file chunks.

**Comment by fahhem** [comment]
The way normal fastboot flash works is to send 2 commands:

* `download:%(size)d` followed by `size` bytes
* `flash:%(partition)s`

From what I gather with the code, they do this instead:

* `getvar max-download-size` to know the 'sparse limit'
* `download:%(total_size)d` followed by the data in up to `max-download-size` byte chunks
* `flash:%(partition)s`

However, the major factor here is what the data is in the second case. There are a bunch of calls into `libsparse` to generate the sparse image. If we're lucky, we can get that data and then send it over correctly. The code that seem relevant here is [sparse's img2simg.c](https://android.googlesource.com/platform/system/core/+/master/libsparse/img2simg.c) which may do it for us, though here is the [initial sparse-file entry point in fastboot](https://android.googlesource.com/platform/system/core.git/+/master/fastboot/fastboot.cpp#641)

What your code is doing is overwriting the data that will be flashed with the latest chunk, so you will end up with the end of your bootloader flashed to the front of the partition.

**Comment by fahhem** [comment]
After looking into it and following the winding codebase, it seems like android-fastboot essentially does this:

* Call `getvar max-download-size` to get the target's max size.
* If the result (`max_size`) is greater than the filesize, then just flash it normally.
* Call `s = sparse_file_import_auto(fd, check_crc=false, verbose=true)` on the file descriptor
* Call `count = sparse_file_resparse(s, max_size, out_s=nullptr, count=0)` to get the size of `out_s`
* Allocate `out_s` as a list of `struct sparse_file` of size `count`
* Call `sparse_file_response(s, max_size, out_s, count)`
  * Now the file is in a list of `struct sparse_file`s called `out_s`. 
* Rip them into a list of (`sparse_file`, block size) pairs by calling `sparse_file_len(s, sparse=true, crc=false)`
* For each pair:
  * Immediately forget that we got the size and get the size again (ugh...) with `sparse_file_len(s, sparse=true, crc=false)`
  * Send a `download:%(block_size)s` with that block size
  * Call `sparse_file_callback(s, sparse=true, crc=false, write=callback, priv=transport)` which calls the callback as `callback(transport, data, len)`
  * In `callback`, concatenate the data and send it in chunks of 1024 bytes
  * Send the last chunk, if any, of bytes (< 1024 by definition)
  * Send a `flash:%(partition)s` for the current partition

So that's how it's done. It looks like `libsparse` is pulled out into a separate `deb` file in newer Debian and Ubuntu distributions, but it seems we'll need to ship a copy internally for now.
